### PR TITLE
FEATURE: BCO-1119 Introduce data sources API

### PR DESCRIPTION
**Pull request by hung-nguyen-hoang**

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/531/ : SUCCESS in 6s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/529/ : SUCCESS in 20m 24s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/533/ : SUCCESS in 5s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/531/ : SUCCESS in 20m 35s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/534/ : SUCCESS in 4s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/532/ : FAILURE in 2m 08s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/535/ : SUCCESS in 4s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/533/ : FAILURE in 4m 14s

**Comment by hung-nguyen-hoang** [comment]
extended test

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/536/ : SUCCESS in 4s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/534/ : SUCCESS in 20m 48s

File: lib/gooddata/models/data_source.rb [hunk 1]
```diff
@@ -0,0 +1,630 @@
...
+      uri.split('/')[-1]
+    end
+
+    def is(type)
+      @json['dataSource']['connectionInfo'][type]
+    end
+
+    private
+
+    def build_connection_info
```
Comment (sangtm) [inline lib/gooddata/models/data_source.rb:161]: Should we have connection_info to get the connection_info?

File: lib/gooddata/models/data_source.rb [hunk 2]
```diff
@@ -0,0 +1,630 @@
...
+        if id == :all
+          data = c.get(DATA_SOURCES_URL)
+          data['dataSources']['items'].map do |ds_data|
+            c.create(DataSource, ds_data)
+          end
+        else
+          c.create(DataSource, c.get("#{DATA_SOURCES_URL}/#{id}"))
+        end
+      end
+
```
[the remark this answers is not in the document]
Comment (hung-nguyen-hoang) [reply lib/gooddata/models/data_source.rb:39]: Done add Docs for all api

File: lib/gooddata/models/data_source.rb [hunk 3]
```diff
@@ -0,0 +1,630 @@
...
+    DATA_SOURCES_URL = '/gdc/dataload/dataSources'
+    SNOWFLAKE = 'snowflake'
+    REDSHIFT = 'redshift'
+    BIGQUERY = 'bigQuery'
+    GENERIC = 'generic'
+    S3 = 's3'
+    ADS = 'ads'
+
+    class << self
+      def [](id = :all, options = { client: GoodData.client })
```
[the remark this answers is not in the document]
Comment (hung-nguyen-hoang) [reply lib/gooddata/models/data_source.rb:27]: In real case, Users rarely have more than 100 data sources. So I think that We don't need provide opt to get more for this.

File: lib/gooddata/models/data_source.rb [hunk 4]
```diff
@@ -0,0 +1,630 @@
...
+          'dataSource' => {
+            'name' => ds_name,
+            'connectionInfo' => ds_connection_info
+          }
+        }
+        json['dataSource']['alias'] = ds_alias if ds_alias
+        json['dataSource']['prefix'] = ds_prefix if ds_prefix
+
+        # Create data source
+        c = GoodData.get_client(opts)
```
[the remark this answers is not in the document]
Comment (hung-nguyen-hoang) [reply lib/gooddata/models/data_source.rb:86]: The opts should contain connection client. If don't contain it, create data source will be failed with error.

File: lib/gooddata/models/data_source.rb (hunk 1 above)
Comment (hung-nguyen-hoang) [reply lib/gooddata/models/data_source.rb:169]: We already have declare "attr_accessor :connection_info" so We can get or set connection_info.

File: lib/gooddata/models/data_source.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (sangtm) [reply lib/gooddata/models/data_source.rb:39]: More detail please: params, return, specical case for input params,... thanks

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/541/ : SUCCESS in 6s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/538/ : SUCCESS in 22m 12s

File: lib/gooddata/models/data_source.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (sangtm) [reply lib/gooddata/models/data_source.rb:39]: For example: before calling save, we have call get or set properties,....

[the remark this answers is not in the document]
Comment (hung-nguyen-hoang) [reply lib/gooddata/models/data_source.rb:39]: Done add more docs for the api. Thank you.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/542/ : SUCCESS in 4s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/539/ : SUCCESS in 22m 41s

**Comment by sangtm** [comment]
LGTM

File: lib/gooddata/models/data_source.rb [hunk 5]
```diff
@@ -0,0 +1,648 @@
...
+      def from_id(id, options = { client: GoodData.client })
+        DataSource[id, options]
+      end
+
+      # Get a specify data source from data source alias
+      def from_alias(data_source_alias, options = { client: GoodData.client })
+        data_sources = all(options)
+        data_sources.find do |data_source|
+          data_source.alias == data_source_alias
+        end
```
Comment (danh-ung) [inline lib/gooddata/models/data_source.rb:50]: What return if id or alias not found, throw exception or return undefined? We should mention in the api doc.

File: lib/gooddata/models/data_source.rb [hunk 6]
```diff
@@ -0,0 +1,648 @@
...
+        # Create data source
+        c = GoodData.get_client(opts)
+        res = c.post(DATA_SOURCES_URL, json)
+
+        # create the public facing object
+        c.create(DataSource, res)
+      end
+
+      # Check whether the data source connection is supported
+      def support_data_source(connection_info_json)
```
Comment (danh-ung) [inline lib/gooddata/models/data_source.rb:94]: Is this functions necessary? When backend supports new data source type, we will update SDK? I think we should rely on the validation on server side.

File: lib/gooddata/models/data_source.rb [hunk 7]
```diff
@@ -0,0 +1,648 @@
...
+
+    private
+
+    def build_connection_info
+      return snowflake_connection_info if is(SNOWFLAKE)
+      return redshift_connection_info if is(REDSHIFT)
+      return bigquery_connection_info if is(BIGQUERY)
+      return generic_connection_info if is(GENERIC)
+      return s3_connection_info if is(S3)
+      return ads_connection_info if is(ADS)
```
Comment (danh-ung) [inline lib/gooddata/models/data_source.rb:185]: Again, how SDK support new data source types?

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/543/ : SUCCESS in 4s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/540/ : SUCCESS in 21m 08s

File: lib/gooddata/models/data_source.rb (hunk 5 above)
Comment (hung-nguyen-hoang) [reply lib/gooddata/models/data_source.rb:56]: Done update api doc. Throw exception if the id or alias are not found.

File: lib/gooddata/models/data_source.rb (hunk 6 above)
Comment (hung-nguyen-hoang) [reply lib/gooddata/models/data_source.rb]: Remove the function. We will rely on the validation on server side.

File: lib/gooddata/models/data_source.rb (hunk 7 above)
Comment (hung-nguyen-hoang) [reply lib/gooddata/models/data_source.rb:186]: Done update code for supporting new data source types.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/544/ : SUCCESS in 5s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/541/ : SUCCESS in 19m 32s

**Review by sangtm — APPROVED** [review]

**Comment by yenkins** [comment]
### Sonar scan result
 - No new bug found. Congratulation!

More detail, see in https://sonarqube-dev.intgdc.com/dashboard?id=gooddata-ruby-gate-PR1714

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-sonarqube-scan-zuul-el7/10/ : SUCCESS in 2m 07s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-merge-compliance-check/487/ : SUCCESS in 21s
- https://checklist.intgdc.com/job/lcm/job/gooddata-ruby-travis-check/136/ : SUCCESS in 20s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-fossa-scanning-zuul-el7/51/ : SUCCESS in 32s (non-voting)

**Pull request merged**